Storage/src/SigningHelper.php

```diff
@@ -0,0 +1,578 @@
+        if ($expires > $expireLimit) {
+            throw new \InvalidArgumentException(
+                'V4 Signed URLs may not have an expiration greater than seven days in the future.'
+            );
+        }
+
+        $clientEmail = $credentials->getClientName();
+        $credentialScope = sprintf('%s/auto/storage/goog4_request', $requestDatestamp);
+        $credential = sprintf('%s/%s', $clientEmail, $credentialScope);
+
+        // Add headers and query params based on provided options.
+        $params = $options['queryParams'];
+        $headers = $options['headers'] + [
+            'host' => $options['cname']
+        ];
+
+        if ($options['contentType']) {
+            $headers['content-type'] = $options['contentType'];
+        }
+
+        if ($options['contentMd5']) {
+            $headers['content-md5'] = $options['contentMd5'];
+        }
+
+        if ($options['responseType']) {
+            $params['response-content-type'] = $options['responseType'];
+        }
+
+        if ($options['responseDisposition']) {
+            $params['response-content-disposition'] = $options['responseDisposition'];
+        } elseif ($options['saveAsName']) {
+            $params['response-content-disposition'] = 'attachment; filename='
+                . '"' . $options['saveAsName'] . '"';
+        }
+
+        if ($generation) {
+            $params['generation'] = $generation;
+        }
+
+        $headers = $this->normalizeHeaders($headers);
+
+        // sort headers by name
+        ksort($headers, SORT_NATURAL | SORT_FLAG_CASE);
+
+        // Canonical headers are a list, newline separated, of keys and values,
+        // comma separated.
+        // Signed headers are a list of keys, separated by a semicolon.
+        $canonicalHeaders = [];
+        $signedHeaders = [];
+        foreach ($headers as $key => $val) {
+            $canonicalHeaders[] = sprintf('%s:%s', $key, $val);
+            $signedHeaders[] = $key;
+        }
+        $canonicalHeaders = implode("\n", $canonicalHeaders) . "\n";
+
+        $signedHeaders = implode(';', $signedHeaders);
+
+        // Add required query parameters.
+        $params['X-Goog-Algorithm'] = self::V4_ALGO_NAME;
+        $params['X-Goog-Credential'] = $credential;
+        $params['X-Goog-Date'] = $requestTimestamp;
+        $params['X-Goog-Expires'] = $expires - $timeSeconds;
+        $params['X-Goog-SignedHeaders'] = $signedHeaders;
+
+        // Sort query string params by name.
+        ksort($params, SORT_NATURAL | SORT_FLAG_CASE);
+
+        // Create a query string, encoding spaces as `%20` rather than `+`.
+        $canonicalQueryString = http_build_query($params, null, '&', PHP_QUERY_RFC3986);
+
+        $canonicalRequest = [
+            $options['method'],
+            $resource,
+            $canonicalQueryString,
+            $canonicalHeaders,
+            $signedHeaders,
+            'UNSIGNED-PAYLOAD'
+        ];
+
+        // Create a request hash to be signed.
```

Comment:
I'm not sure we need the comment here or on some of the other lines below, the code reads pretty clearly.